pdns/recursordist/pdns_recursor.cc

```diff
@@ -267,9 +267,23 @@ static void handleUDPServerResponse(int fileDesc, FDMultiplexer::funcparam_t& va
 
 thread_local std::unique_ptr<UDPClientSocks> t_udpclientsocks;
 
+// If we have plenty of mthreads slot left, use default timeout.
+// Otherwise reduce the timeout to be between g_networkTimeoutMsec/10 and g_networkTimeoutMsec
+unsigned int authWaitTimeMSec(const std::unique_ptr<MT_t>& mtasker)
+{
+  const auto max = g_maxMThreads;
+  const auto current = mtasker->numProcesses();
+  const unsigned int cutoff = max / 10; // if we have less than 10% used,  do not reduce auth timeout
+  if (current < cutoff) {
+    return g_networkTimeoutMsec;
+  }
+  const auto avail = max - current;
+  return std::max(g_networkTimeoutMsec / 10, g_networkTimeoutMsec * avail / (max - cutoff));
```

Comment:
One thing that bothers me is that we do throttle authoritative servers on a timeout, but it doesn't seem fair to treat a N seconds timeout the same than a N/10 seconds one. Do you think it would make sense not to throttle if the timeout was below a certain threshold (80 % of `g_networkTimeoutMsec`, or something like that)?

Reply:
Agreed, will look into that.